drivers/input/touchscreen/himax.c

```diff
@@ -0,0 +1,381 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Driver for Himax touchscreens
+ *
+ * Copyright (C) 2022 Job Noorman <job@noorman.info>
+ *
+ * This code is based on "Himax Android Driver Sample Code for QCT platform":
+ *
+ * Copyright (C) 2017 Himax Corporation.
+ */
+
+#include <asm/byteorder.h>
+#include <linux/delay.h>
+#include <linux/err.h>
+#include <linux/gpio/consumer.h>
+#include <linux/i2c.h>
+#include <linux/input.h>
+#include <linux/input/mt.h>
+#include <linux/input/touchscreen.h>
+#include <linux/interrupt.h>
+#include <linux/kernel.h>
+#include <linux/regmap.h>
+
+#define HIMAX_ID_83112A 0x83112a
+#define HIMAX_ID_83112B 0x83112b
+
+#define HIMAX_MAX_POINTS 10
+
+#define HIMAX_REG_CFG_SET_ADDR 0x00
+#define HIMAX_REG_CFG_INIT_READ 0x0c
+#define HIMAX_REG_CFG_READ_VALUE 0x08
+#define HIMAX_REG_READ_EVENT 0x30
+
+#define HIMAX_CFG_PRODUCT_ID 0x900000d0
+
+struct himax_event_point {
+	__be16 x;
+	__be16 y;
+} __packed;
+
+struct himax_event {
+	struct himax_event_point points[HIMAX_MAX_POINTS];
+	u8 majors[HIMAX_MAX_POINTS];
+	u8 pad0[2];
+	u8 num_points;
+	u8 pad1[2];
+	u8 checksum_fix;
+} __packed;
+
+static_assert(sizeof(struct himax_event) == 56);
+
+struct himax_ts_data {
+	struct gpio_desc *gpiod_rst;
+	struct input_dev *input_dev;
+	struct i2c_client *client;
+	struct regmap *regmap;
+	struct touchscreen_properties props;
+};
+
+static const struct regmap_config himax_regmap_config = {
+	.reg_bits = 8,
+	.val_bits = 32,
+	.val_format_endian = REGMAP_ENDIAN_LITTLE,
+};
+
+static int himax_read_config(struct himax_ts_data *ts, u32 address, u32 *dst)
+{
+	int error = 0;
+
+	error = regmap_write(ts->regmap, HIMAX_REG_CFG_SET_ADDR, address);
+	if (error)
+		return error;
+
+	error = regmap_write(ts->regmap, HIMAX_REG_CFG_INIT_READ, 0x0);
+	if (error)
+		return error;
+
+	return regmap_read(ts->regmap, HIMAX_REG_CFG_READ_VALUE, dst);
+}
+
+static int himax_read_input_event(struct himax_ts_data *ts,
+				  struct himax_event *event)
+{
+	return regmap_raw_read(ts->regmap, HIMAX_REG_READ_EVENT, event,
+			       sizeof(*event));
+}
+
+static void himax_reset(struct himax_ts_data *ts)
+{
+	gpiod_set_value(ts->gpiod_rst, 1);
+	msleep(20);
+	gpiod_set_value(ts->gpiod_rst, 0);
+}
```

Comment:
I have noticed that after a handful of `rmmod`/`modprobe` cycles, probing the driver often fails due to a failed I2C transfer (`-EIO`). I suppose this could have something to do with this reset sequence. Or could you see another reason for this?

Reply:
Actually, I don't seem to be able to reproduce this anymore. I encountered it a couple of while developing after `scp`ing a new module to the phone. When I don't change the module and just call `rmmod himax && modprobe himax` in a loop, it seems to keep working just fine.

Reply:
Ah, I know how to reproduce it: it happens when the screen is turned off. I guess that makes sense as the touchscreen controller shares a regulator with the panel?

Reply:
> Ah, I know how to reproduce it: it happens when the screen is turned off. I guess that makes sense as the touchscreen controller shares a regulator with the panel?

@mtvec This is why you do need regulator handling (including in suspend/resume together with reset);

Reply:
I just tried this and it doesn't make a difference. I might be using the wrong regulators though. I'm using:

```
vcc-i2c-supply = <&pm8953_l6>;
vdd-ana-supply = <&pm8953_l10>;
```

I'm trying to figure out which regulators I should use by running the following command:

```bash
for r in /sys/class/regulator/*; do echo $r; cat $r/name; cat $r/state; done
```

I ran this once while the screen is on and once while it is off  right after enabling the regulators in the driver's probe. By diffing the outputs, I see there is one regulator that is enabled in the first case, but disabled in the second:

```
 /sys/class/regulator/regulator.32
 s3
-enabled
+disabled
```

So I seem to need `s3`. By running the same command while the screen is on, I can also see `l6` changes state but `l10` does not. So I updated the dts like so:

```
vdd-ana-supply = <&pm8953_s3>;
```

But this doesn't make any difference.

One interesting thing I noticed is this:

```bash
$ cat /sys/class/regulator/regulator.42/name
l6
$ cat /sys/class/regulator/regulator.42/num_users
1
$ sudo modprobe himax
$ cat /sys/class/regulator/regulator.42/num_users
3
```

So for some reason the driver uses `l6` *twice*.

It feels like there is something wrong with how the regulators are setup in the dts. Any ideas?

(If you're interested, the code is [here](https://github.com/mtvec/msm8953-mainline-linux/commit/da84614a144aa9ebae6b4822c9af68762e655715))

Reply:
Possible that the display+touchscreen component just disables touch functionality when screen is turned off, and I also don't think it's a necessary use case that the touchscreen works when the screen is off. Downstream also can't do this.